### PR TITLE
[no sq] Lua JSON util improvements

### PR DESCRIPTION
**Pull request by appgurueu**
Currently modders can't handle JSON parsing errors themselves, we always log an error message. This PR changes that.

I'm not sure whether to consider this a bugfix or a feature. Personally I consider the "always log" behavior more of a quirk / bug but technically this PR might be a feature, since the previous behavior was documented and working as intended.

Also added another commit which fixes the low max stack depth for reading and makes the depths for reading and writing consistent.

## How to test

See the unit test.

### CODE REVIEW

File: src/script/lua_api/l_util.cpp [hunk 1]
```diff
@@ -131,9 +138,15 @@ int ModApiUtil::l_parse_json(lua_State *L)
...
-		errorstream << "Failed to parse json data, "
-			<< "depth exceeds lua stack limit" << std::endl;
+		const char *err_stack_depth = "depth exceeds lua stack limit";
+		if (return_error) {
+			lua_pushnil(L);
+			lua_pushstring(L, err_stack_depth);
+			return 2;
+		}
+		errorstream << "Failed to parse json data, " << err_stack_depth << std::endl;
 		errorstream << "data: \"" << jsonstr << "\"" << std::endl;
```
Comment (sfan5) [inline src/script/lua_api/l_util.cpp:150]: this should be made consistent with the other one.
tho i don't see the point in this errorstream/warningstream distinction, so you can get rid of that.

Comment (appgurueu) [reply src/script/lua_api/l_util.cpp]: sure. done.

File: games/devtest/mods/unittests/misc.lua [hunk 2]
```diff
@@ -155,13 +155,22 @@ unittests.register("test_urlencode", test_urlencode)
...
+	end
+	do
+		local null = {}
+		local data = core.parse_json(raw, null)
+		assert(data.z == null)
+	end
+	do
+		local data, err = core.parse_json('"ceci n\'est pas un json', nil, true)
+		assert(data == nil)
+		assert(err ~= nil)
```
Comment (sfan5) [inline games/devtest/mods/unittests/misc.lua:172]:
```suggestion
		assert(type(err) == "string")
```

**Review by sfan5 — APPROVED** [review]
lgtm

**Comment by appgurueu** [comment]
Also got rid of two unnecessary string copies by directly giving a string view to jsoncpp. (Unfortunately this doesn't improve performance much because jsoncpp is still relatively wasteful.)

**Review by sfan5 — APPROVED** [review]
lgtm if tests pass

**Comment by appgurueu** [comment]
Added another commit for a max depth of 1024 for writing and reading

File: src/script/lua_api/l_util.cpp [hunk 3]
```diff
@@ -93,12 +93,16 @@ int ModApiUtil::l_get_us_time(lua_State *L)
 	return 1;
 }
 
-// parse_json(str[, nullvalue])
+// Maximum depth of a JSON object:
+// Reading and writing should not overflow the Lua, C, or jsoncpp stacks.
+const u16 MAX_JSON_DEPTH = 1024;
```
Comment (sfan5) [inline src/script/lua_api/l_util.cpp:98]: should be `const static` or `constexpr static`

**Review by sfan5 — APPROVED** [review]
lgtm

**Pull request merged**